lgc/patch/LowerDebugPrintf.cpp

```diff
@@ -0,0 +1,257 @@
+    auto name = func.getName();
+    if (name.startswith(lgcName::LowerDebugPrintf)) {
+      for (auto user : func.users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          builder.SetInsertPoint(callInst);
+          auto resultVal = createDebugPrintf(callInst->getArgOperand(0), callInst->getArgOperand(1),
+                                             make_range(callInst->arg_begin() + 2, callInst->arg_end()), builder);
+          callInst->replaceAllUsesWith(resultVal);
+          callees.push_back(callInst);
+        }
+      }
+    }
+  }
+
+  for (auto callInst : callees) {
+    callInst->eraseFromParent();
+  }
+
+  if (m_elfInfos.size() == 0)
+    return PreservedAnalyses::none();
+
+  setupElfsPrintfStrings();
+  return PreservedAnalyses::all();
+}
+
+// =====================================================================================================================
+// Create debug printf operation, and write to the output debug buffer
+// @debugPrintfBuffer : Output buffer for debug print data
+// @formatStr : Printf format string
+// @vars : Printf variable parameters
+// @builder : BuilderBase to build instruction
+llvm::Value *LowerDebugPrintf::createDebugPrintf(Value *debugPrintfBuffer, Value *formatStr,
+                                                 iterator_range<User::op_iterator> vars, BuilderBase &builder) {
+
+  // Printf output variables in DWORDs
+  SmallVector<Value *> printArgs;
+  // Records printf output variables are 64bit or not
+  SmallBitVector bit64Vector;
+  for (const auto &var : vars) {
+    getDwordValues(var, printArgs, bit64Vector);
+  }
+
+  GlobalVariable *globalStr = cast<GlobalVariable>(formatStr);
+  StringRef strDebugStr = (cast<ConstantDataSequential>(globalStr->getInitializer()))->getAsString();
+
+  uint64_t hash = hash_value(strDebugStr);
+
+  static const unsigned EntryHeaderSize = 2; // 2Dword EntrySize + stringId
+  unsigned entrySize = EntryHeaderSize + printArgs.size();
+
+  // 64 bit header {[0:15], [16:63]} entrySize,hash value for the string
+  uint64_t header = (hash << 16) | entrySize;
+  // save 48bit hash value to the m_formatStrings
+  uint64_t hashValue = header >> 16;
+  m_elfInfos[hashValue].formatString = strDebugStr;
+  m_elfInfos[hashValue].bit64Pos = bit64Vector;
+
+  uint32_t loEntryheader = uint32_t(header);
+  uint32_t hiEntryheader = uint32_t(header >> 32);
+
+  auto offsetTy = PointerType::get(builder.getInt64Ty(), ADDR_SPACE_BUFFER_FAT_POINTER);
+  Value *bufferPtr = builder.CreateBitCast(debugPrintfBuffer, offsetTy);
+  // uint64_t offset = AtomicAdd64(offsetPtr, entrySize);
+  // maxOffset = 1<<31;
+  // offset = offset < maxOffset ? offset : maxOffset;
+  Value *entryOffset = builder.CreateAtomicRMW(AtomicRMWInst::Add, bufferPtr, builder.getInt64(entrySize), MaybeAlign(),
+                                               AtomicOrdering::Monotonic, SyncScope::SingleThread);
+  Value *maxOffset = builder.getInt64(1 << 31);
```

Comment:
`1 << 31` is taken as a signed int and is sign-extended to an uint64_t.
The CsLowerDebugPrintf.lgc test shows that it ends up as `-2147483648 = 0xffffffff80000000`, but it should be `0x80000000`.